runtime/src/main/kotlin/kotlin/reflect/AssociatedObjects.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.reflect
+
+import kotlin.native.internal.*
+
+/**
+ * The experimental marker for associated objects API.
+ *
+ * Any usage of a declaration annotated with `@ExperimentalAssociatedObjects` must be accepted either by
+ * annotating that usage with the [UseExperimental] annotation, e.g. `@UseExperimental(ExperimentalAssociatedObjects::class)`,
+ * or by using the compiler argument `-Xuse-experimental=kotlin.reflect.ExperimentalAssociatedObjects`.
+ */
+@Experimental(level = Experimental.Level.ERROR)
+@Retention(value = AnnotationRetention.BINARY)
+public annotation class ExperimentalAssociatedObjects
+
+/**
+ * Makes the annotated annotation class an associated object key.
+ *
+ * An associated object key annotation should have single [KClass] parameter.
+ * When applied to a class with reference to an object declaration as an argument, it binds
+ * the object to the class, making this binding discoverable at runtime using [findAssociatedObject].
+ */
+@ExperimentalAssociatedObjects
+@Retention(AnnotationRetention.BINARY)
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+public annotation class AssociatedObjectKey
+
+/**
+ * If [T] is an @[AssociatedObjectKey]-annotated annotation class and [this] class is annotatated with @[T] (`S::class`),
```

Comment:
Another wording suggestion, change type parameter of key to `K` and then
```
If this class is annotated with an annotation of type [K], which is itself an associated object key (see [AssociatedObjectKey] annotation), 
returns an instance of the class specified by that annotation or `null` otherwise.
```

Reply:
Kinda unreadable:
> an annotation of type [K]

> (see [AssociatedObjectKey] annotation)

> an instance of the class specified by that annotation

which one?

Also I consider my version more self-contained and not relying on `AssociatedObjectKey` documentation that much.

Reply:
I was having hard time deciphering what "@[T] (`S::class`)" means :)
That's the problem inherent to documentation of meta-annotations.

Is this better?
> an instance of the class specified by that `K` annotation

Reply:
> I was having hard time deciphering what "@[T] (S::class)" means :)

It is rendered as `@T (S::class)`. A space character may be confusing but is required for kdoc parser to recognize `[T]` properly. I can avoid using link here to make it `T(S::class)` which is supposed to be pretty straightforward and understandable.

> Is this better?

Still has a couple of issues:

* Relies on  `AssociatedObjectKey` documentation too much.
* "an instance of the class" is too low-level. Kotlin object declarations don't seem to be referred as classes in the official documentation, so "instance of the class" is somewhat unclear from the language point of view.